jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,445 @@
+         *
+         * @return the labels of all vertices of the path
+         */
+        protected Set<Integer> getLabels() {
+            return labels;
+        }
+
+        /**
+         * Returns the cost of the path
+         *
+         * @return the cost of the path
+         */
+        protected double getCost() {
+            return cost;
+        }
+
+        /**
+         * Returns a shallow copy of this labeled path instance. Vertices are not cloned.
+         *
+         * @return a shallow copy of this path.
+         *
+         * @throws RuntimeException in case the clone is not supported
+         *
+         * @see java.lang.Object#clone()
+         */
+        public LabeledPath<V> clone() {
+            try {
+                LabeledPath<V> newLabeledPath = TypeUtil.uncheckedCast(super.clone());
+                newLabeledPath.vertices = (ArrayList<V>) this.vertices.clone();
+                newLabeledPath.labels = (HashSet<Integer>) this.labels.clone();
+                newLabeledPath.cost = this.cost;
+
+                return newLabeledPath;
+            } catch (CloneNotSupportedException e) {
+                e.printStackTrace();
+                throw new RuntimeException();
+            }
+        }
+    }
+
+    /**
+     * Implementation of an key for the path maps.
+     * It is used in AhujaOrlinSharmaCyclicExchangeLocalAugmentation to efficiently maintain the path sets in the calculation.
+     *
+     * @param <V> the vertex type
+     *
+     * @author Christoph Grüne
+     * @since June 7, 2018
+     */
+    private static class PathSetKey<V> {
+        /**
+         * the head of the paths indexed by this key
+         */
+        private V head;
+        /**
+         * the tail of the paths indexed by this key
+         */
+        private V tail;
+        /**
+         * the label set of the paths indexed by this key
+         */
+        private Set<Integer> labels;
+
+        /**
+         * Constructs a new PathSetKey object
+         *
+         * @param head the head of the paths indexed by this key
+         * @param tail the tail of the paths indexed by this key
+         * @param labels the label set of the paths indexed by this key
+         */
+        private PathSetKey(V head, V tail, Set<Integer> labels) {
+            this.head = head;
+            this.tail = tail;
+            this.labels = labels;
+        }
+
+        /**
+         * Constructs a new PathSetKey object
+         *
+         * @param head the head of the path indexed by this key
+         * @param tail the tail of the path indexed by this key
+         * @param labels the label set of the path indexed by this key
+         * @param <V> the vertex type
+         * @return a new PathSetKey object
+         */
+        protected static <V> PathSetKey<V> of(V head, V tail, Set<Integer> labels) {
+            return new PathSetKey<>(head, tail, labels);
+        }
+
+        @Override
+        public int hashCode() {
```

Comment:
It's cleaner to use Objects.hash

https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...)